feat: add C ndarray interface and refactor implementation for `stats/base/dstdev`

0PrashantYadav0 opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds C ndarray interface for stats/base/dstdev
-   refactors JavaScript implementation
-   updates tests, docs, examples and benchmarks

## Related Issues

> Does this pull request have any related issues?

NO

-   resolves #{{TODO: add issue number}} 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

0PrashantYadav0 committed:
feat: add C ndarray interface and refactor implementation

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 committed:
chore: updated main.c file

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dstdev/README.md:
```diff
@@ -224,6 +208,125 @@ console.log( v );
+const double x[] = { 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 };
+
+double v = stdlib_strided_dstdev_ndarray( 4, 1.0, x, 2, 1 );
+// returns ~2.0817
```
@0PrashantYadav0 
Hey, shouldn't the return value be 2.5 as:-
The subset elements are:- { 1.0, -2.0, 2.0, 4.0 } 
Mean = ( 1.0 + (-2.0) + 2.0 + 4.0 )/ 4 =1.25
Deviations form the mean:-
1) 0.0625 2) 10.5625 3) 0.5625 4) 7.5625
Sum of squared deviations = 0.0625+10.5625+0.5625+7.5625=18.75
Will be dividing it by:- 
𝑁−1=4−1=3 (as correction is 1.0):
Variance = 18.75/3 = 6.25
stdev= sqrt(6.25) = 2.5

0PrashantYadav0 replied:
Sorry i change that by mistake

Neerajpathak07 replied:
Might want to make these changes through out your other PR's related to similar package

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dstdev/src/main.c:
```diff
@@ -0,0 +1,50 @@
+* @return             output value
+*/
+double API_SUFFIX(stdlib_strided_dstdev_ndarray)( const CBLAS_INT N, const double correction, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
+	return API_SUFFIX(stdlib_strided_dstdevpn_ndarray)( N, correction, X, strideX, offsetX );
```
also you don't need `dvariancepn` instead you need C implementation for `dstdevpn`.
I'm not sure if you have created a PR for that but just wanted to give a heads up!

0PrashantYadav0 replied:
Yes i have created a PR for that

0PrashantYadav0 committed:
fix: changes return values

0PrashantYadav0 committed:
fix: changes return values

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

0PrashantYadav0 commented:
Please merge #4688 PR and then #4687 PR before merging this.

0PrashantYadav0 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into ndarray-dstdev

0PrashantYadav0 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into ndarray-dstdev

0PrashantYadav0 committed:
chore: merge to develop branch

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

0PrashantYadav0 committed:
chore: minor changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

0PrashantYadav0 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into ndarray-dstdev

0PrashantYadav0 committed:
fix: fix lint error and import statements

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

0PrashantYadav0 commented:
@kgryte ready for review.

kgryte commented on lib/node_modules/@stdlib/stats/base/dstdev/README.md:
```diff
@@ -98,17 +98,15 @@ The use of the term `n-1` is commonly referred to as Bessel's correction. Note,
-#### dstdev( N, correction, x, stride )
+#### dstdev( N, correction, x, strideX )
 
 Computes the [standard deviation][standard-deviation] of a double-precision floating-point strided array `x`.
```
```suggestion
Computes the [standard deviation][standard-deviation] of a double-precision floating-point strided array.
```

kgryte committed:
docs: update description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/dstdev/README.md:
```diff
@@ -138,45 +134,39 @@ Note that indexing is relative to the first index. To introduce an offset, use [
+-   **offsetX**: starting index for `x`.
 
-While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on a starting index. For example, to calculate the [standard deviation][standard-deviation] for every other value in `x` starting from the second value
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on a starting index. For example, to calculate the [standard deviation][standard-deviation] for every other element in `x` starting from the second element
```
```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter supports indexing semantics based on a starting index. For example, to calculate the [standard deviation][standard-deviation] for every other element in `x` starting from the second element
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/dstdev/README.md:
```diff
@@ -224,6 +208,125 @@ console.log( v );
+
+#### stdlib_strided_dstdev( N, correction, \*X, strideX )
+
+Computes the [standard deviation][standard-deviation] of a double-precision floating-point strided array `x`.
```
```suggestion
Computes the [standard deviation][standard-deviation] of a double-precision floating-point strided array.
```

kgryte committed:
docs: update description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/dstdev/README.md:
```diff
@@ -224,6 +208,125 @@ console.log( v );
+const double x[] = { 1.0, -2.0, 2.0 };
+
+double v = stdlib_strided_dstdev( 3, 1.0, x, 1 );
+// returns 2.5
```
This isn't correct.

@0PrashantYadav0 Please go through this PR and more closely determine whether further things need updated.

0PrashantYadav0 replied:
```suggestion
// returns ~2.0817
```

0PrashantYadav0 commented on lib/node_modules/@stdlib/stats/base/dstdev/README.md:
```diff
@@ -224,6 +208,125 @@ console.log( v );
+
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **correction**: `[in] double` degrees of freedom adjustment. Setting this parameter to a value other than `0` has the effect of adjusting the divisor during the calculation of the [standard deviation][standard-deviation] according to `N-c` where `c` corresponds to the provided degrees of freedom adjustment. When computing the [standard deviation][standard-deviation] of a population, setting t ...
+-   **x**: `[in] double*` input array.
```
```suggestion
-   **X**: `[in] double*` input array.
```

0PrashantYadav0 commented on lib/node_modules/@stdlib/stats/base/dstdev/lib/index.js:
```diff
@@ -39,9 +37,7 @@
 * var dstdev = require( '@stdlib/stats/base/dstdev' );
```
remove floor import statement

0PrashantYadav0 commented on lib/node_modules/@stdlib/stats/base/dstdev/manifest.json:
```diff
@@ -27,17 +28,19 @@
       "dependencies": [
+        "@stdlib/blas/base/shared",
+        "@stdlib/strided/base/stride2offset",
+        "@stdlib/math/base/special/sqrt",
```
remove unwanted dependencies

0PrashantYadav0 commented on lib/node_modules/@stdlib/stats/base/dstdev/src/addon.c:
```diff
@@ -35,11 +36,29 @@
-	STDLIB_NAPI_CREATE_DOUBLE( env, stdlib_strided_dstdev( N, correction, X, stride ), v );
+	STDLIB_NAPI_ARGV_INT64( env, strideX, argv, 3 );
+	STDLIB_NAPI_ARGV_STRIDED_FLOAT64ARRAY( env, X, N, strideX, argv, 2 )
+	STDLIB_NAPI_CREATE_DOUBLE( env, API_SUFFIX(stdlib_strided_dstdev)( N, correction, X, strideX ), v );
```
```suggestion
	STDLIB_NAPI_CREATE_DOUBLE( env, API_SUFFIX( stdlib_strided_dstdev )( N, correction, X, strideX ), v );
```

kgryte replied:
This isn't correct. Within the `API_SUFFIX` macro, we haven't been including spaces. See other packages.

0PrashantYadav0 replied:
Okay i will not change anything here. Thanks.

0PrashantYadav0 commented on lib/node_modules/@stdlib/stats/base/dstdev/src/addon.c:
```diff
@@ -35,11 +36,29 @@
+	STDLIB_NAPI_ARGV_INT64( env, strideX, argv, 3 );
+	STDLIB_NAPI_ARGV_STRIDED_FLOAT64ARRAY( env, X, N, strideX, argv, 2 );
+	STDLIB_NAPI_ARGV_INT64( env, offsetX, argv, 4 );
+	STDLIB_NAPI_CREATE_DOUBLE( env, API_SUFFIX(stdlib_strided_dstdev_ndarray)( N, correction, X, strideX, offsetX ), v );
```
```suggestion
	STDLIB_NAPI_CREATE_DOUBLE( env, API_SUFFIX( stdlib_strided_dstdev_ndarray )( N, correction, X, strideX, offsetX ), v );
```

kgryte replied:
Same thing.

0PrashantYadav0 replied:
Okay.

0PrashantYadav0 committed:
chore: changes as per code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

0PrashantYadav0 commented:
@kgryte Ready for review again.

kgryte commented:
@0PrashantYadav0 Would you mind resolving the merge conflicts?

kgryte commented:
Ref: https://github.com/stdlib-js/stdlib/pull/6850
